### ISSUE TITLE
Delimiters (e.g. O', Mc, -) don't work as expected

### ISSUE THREAD

**Issue by MegaphoneJon** [cividesk/com.cividesk.normalize#4]
The code to handle delimiters when capitalizing names in CRM_Utils_Normalize::normalize_contact is cribbed from here: http://php.net/manual/en/function.ucwords.php#112795

However, this code is fundamentally broken.  The foreach loop ensures that only the last delimiter in the $delimiters array matters.

I've tried thinking of ways to fix this code without changing the general approach - but I think the only real fix is to employ a clever preg_replace.

**Comment by nganivet** [comment]
Jon - yes, we really struggled with all the exceptions, specially as we 
have international customers and rules are not always the same in 
different parts of the world. I hate the preg_replace for 
maintainability: a regex is hard enough to correct when you've written 
it yourself, trying to correct a regex written by somebody else is an 
unfathomable challenge. In the end I am thinking we might be better off 
managing these exceptions on an individual basis with a series of ifs.

ie.

if first two letters in ("O'", "Mc") then capitalize 3rd
if <another test> then <another action)
[...]

------ Original Message ------

**Comment by PalanteJon** [comment]
Hi Nicolas - Your alternative suggestion is a good approach IMO.  I'll talk to my client who requested this fix and see if I can get them to sponsor me doing it.

**Comment by nganivet** [comment]
Super, keep me updated.

------ Original Message ------

**Comment by mlutfy** [comment]
Here's a function I wrote, also ties in fixes from #5 

```
  function ucFirst($string) {
    // https://github.com/cividesk/com.cividesk.normalize/issues/5
    $string = str_replace("’", "'", $string);
    
    // Use mb_convert_case otherwise strings such as MONTRÉAL end up as MontrÉal
    $string = mb_convert_case(mb_strtolower($string), MB_CASE_TITLE, "UTF-8");
    
    // Fix strings such as O'Connor or L'Ancienne-Lorette (city)
    // or names such as McEachran (street name in Montreal).
    // We start by a quick check, to avoid doing the split/implode otherwise.
    if (strpos($string, "'") !== FALSE || strpos($string, 'Mc')) {
      $parts = explode(' ', $string);
        
      foreach ($parts as &$part) {
        if (!empty($part[1]) && $part[1] == "'") {
          // O'connor -> O'Connor
          $t = mb_strtoupper($part[2]);
          $part[2] = $t;
        }
        elseif (!empty($part[2]) && substr($part, 0, 2) == 'Mc') {
          // Mceachran -> McEachran
          $t = mb_strtoupper($part[2]);
          $part[2] = $t;
        }
      }
    
      $string = implode(' ', $parts);
    }
      
    return $string;
  }       
```

**Comment by nganivet** [comment]
@mlutfy Can you please provide a PR for this? Thanks.
